addons/mail/static/tests/owl/hooks/use_store_compare_keys_tests.js

```diff
@@ -0,0 +1,218 @@
+odoo.define('mail.hooks.useStoreCompareKeysTests', function (require) {
+'use strict';
+
+const { proxyComparator, proxyComparatorDeep, useStoreCompareKeys } = require('mail.hooks.useStoreCompareKeys');
+
+const { Component, QWeb, Store } = owl;
+const { xml } = owl.tags;
+
+const {
+    beforeEach: utilsBeforeEach,
+    afterEach: utilsAfterEach,
+    afterNextRender,
+    nextAnimationFrame,
+} = require('mail.messagingTestUtils');
+
+QUnit.module('mail.messaging', {}, function () {
+QUnit.module('hooks', {}, function () {
+QUnit.module('useStoreCompareKeysTests', {
+    beforeEach() {
+        utilsBeforeEach(this);
+        const qweb = new QWeb();
+        this.env = { qweb };
+        this.start = () => {
+            this.store = new Store({
+                env: this.env,
+                state: {
+                    obj: {
+                        a: 1,
+                        b: 2,
+                        c: 3,
+                    },
+                },
+            });
+            this.env.store = this.store;
+            this.hashFn = this.store.observer.revNumber.bind(this.store.observer);
+            this.isEqual = (a, b) => a === b;
+        };
+    },
+    afterEach() {
+        utilsAfterEach(this);
+        this.env = undefined;
+        this.store = undefined;
+        this.hashFn = undefined;
+        this.isEqual = undefined;
+    }
+});
+
+QUnit.test("proxyComparator when strict equal", async function (assert) {
+    assert.expect(6);
+    this.start();
+    const comparator = proxyComparator(this.hashFn, this.isEqual);
+    assert.ok(comparator(1, 1));
+    assert.ok(comparator('a', 'a'));
+    assert.notOk(comparator('a', 'b'));
+    assert.notOk(comparator('1', 1));
+    const a = {};
+    assert.ok(comparator(a, a));
+    assert.notOk(comparator(a, {}));
+});
+
+QUnit.test("proxyComparator on store proxy: temporal comparison", async function (assert) {
+    assert.expect(7);
+    this.start();
+    const comparator = proxyComparator(this.hashFn, this.isEqual);
+    const obj = this.store.state.obj;
+    assert.notOk(comparator(obj, obj), 'initial state of proxy, always different');
+    assert.ok(comparator(obj, obj), 'unchanged proxy');
+    obj.a = 2;
+    assert.notOk(comparator(obj, obj), 'updated proxy after primitive set');
+    assert.ok(comparator(obj, obj), 'unchanged proxy after primitive set');
+    obj.array = [];
+    assert.notOk(comparator(obj, obj), 'updated proxy after add key');
+    obj.array.push(1);
+    assert.notOk(comparator(obj, obj), 'updated proxy after array push');
+    assert.ok(comparator(obj, obj), 'unchanged proxy after array push');
```

Comment:
for readability, could group interaction and assertions together, separated by empty line.
Example:

```js
    const obj = this.store.state.obj;
    assert.notOk(comparator(obj, obj), 'initial state of proxy, always different');
    assert.ok(comparator(obj, obj), 'unchanged proxy');

    obj.a = 2;
    assert.notOk(comparator(obj, obj), 'updated proxy after primitive set');
    assert.ok(comparator(obj, obj), 'unchanged proxy after primitive set');

    obj.array = [];
    assert.notOk(comparator(obj, obj), 'updated proxy after add key');

    obj.array.push(1);
    assert.notOk(comparator(obj, obj), 'updated proxy after array push');
    assert.ok(comparator(obj, obj), 'unchanged proxy after array push');
```

Reply:
ok